src/main.rs

```diff
@@ -47,55 +51,83 @@ fn compile(txt: &str) -> (String, Vec<usize>) {
     (out, map)
 }
 
-fn interp(prg: &[u8], map: Vec<usize>) {
-    let mut pc = 0;
-    let mut ptr = 0;
-    let mut cells = Vec::with_capacity(30000);
-    cells.resize(30000, 0u8);
-    let mut stdin = std::io::stdin().bytes();
-    let mut stdout = std::io::stdout();
-    while pc < prg.len() {
-        match prg[pc] as char {
-            '>' => {
-                if ptr == cells.len() {
-                    cells.push(0);
-                }
-                ptr += 1;
-            }
-            '<' => {
-                if ptr > 0 {
-                    ptr -= 1;
-                }
-            }
-            '+' => {
-                cells[ptr] = cells[ptr].wrapping_add(1);
-            }
-            '-' => {
-                cells[ptr] = cells[ptr].wrapping_sub(1);
+#[derive(Debug)]
+struct InterpCx {
+    pc: usize,
+    ptr: usize,
+    cells: Vec<u8>,
+    prog: Vec<u8>,
+    map: Vec<usize>,
+    stdin: Bytes<Stdin>,
+    stdout: Stdout,
+}
+
+fn interp(prog: &[u8], map: Vec<usize>) {
+    let mut icx = InterpCx {
+        pc: 0,
+        ptr: 0,
+        cells: Vec::from([0; 30000]),
+        map,
+        prog: prog.to_vec(),
+        stdin: std::io::stdin().bytes(),
+        stdout: std::io::stdout(),
+    };
+
+    // FIXME only use locations for loop heads.
```

Comment:
When I added the merge point in one of my branches a while ago I did this:
```rust
while tio.pc < prg.len() {
    let loc = if tio.pc > 1 && (tio.prg[tio.pc-1] as char) == '[' {
        Some(&*map.entry(tio.pc).or_insert_with(|| Location::new()))
    }
    else {
        None
    };
    mtt.control_point(loc, interp_inner, &mut tio);
}
```

Reply:
Nice, i'll slot that in.

Reply:
I've implemented this using a hashmap like you did Lukas, however I have an idea to improve this later.